test/apicoverage-webhook/webhook_handlers.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"container/list"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"github.com/knative/test-infra/tools/webhook-apicoverage/coveragecalculator"
+	"github.com/knative/test-infra/tools/webhook-apicoverage/resourcetree"
+	"github.com/knative/test-infra/tools/webhook-apicoverage/view"
+	"go.uber.org/zap"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+
+
+const (
+	// ResourceQueryParam query param name to provide the resource.
+	ResourceQueryParam = "resource"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+// APICoverageRecorder type contains resource tree to record API coverage for resources.
+type APICoverageRecorder struct {
+	Decoder runtime.Decoder
+	Logger *zap.SugaredLogger
+	resourceForest resourcetree.ResourceForest
+}
+
+func (a *APICoverageRecorder) init() {
+	a.resourceForest = resourcetree.ResourceForest {
+		Version: "v1alpha1",
+		ConnectedNodes: make(map[string]*list.List),
+		TopLevelTrees: make(map[string]resourcetree.ResourceTree),
+	}
+	a.addResourceTree("Service", reflect.TypeOf(v1alpha1.Service{}))
+	a.addResourceTree("Configuration", reflect.TypeOf(v1alpha1.Configuration{}))
+	a.addResourceTree("Route", reflect.TypeOf(v1alpha1.Route{}))
+	a.addResourceTree("Revision", reflect.TypeOf(v1alpha1.Revision{}))
+}
+
+func (a *APICoverageRecorder) addResourceTree(resource string, t reflect.Type) {
+	tree := resourcetree.ResourceTree{
+		ResourceName: resource,
+		Forest: &a.resourceForest,
+	}
+	tree.BuildResourceTree(t)
+	a.resourceForest.TopLevelTrees[resource] = tree
+}
+
+func (a *APICoverageRecorder) recordCoverage(review *v1beta1.AdmissionReview) {
+	switch schemaType := review.Request.Kind.Kind; schemaType {
+	case "Service":
```

Comment:
Your 3 cases use the same code. I suggest using a helper function.

Reply:
```
```

	route := v1alpha1.Route{}
		err := json.Unmarshal(review.Request.Object.Raw, &route)
		if err != nil {
			a.Logger.Errorf("Error unmarshalling Service Obj. Request Obj: %s Error: %v", string(review.Request.Object.Raw), err)
			return
		}
```
This piece needs to be specific per case as it depends on the case. I tried json.UnMarshall with ```interface{}``` but it was unable to correctly typecast. Remaining piece of code is refactored.

Reply:
Bummer :/

Reply:
Looks like `reflect.ValueOf(xxx)` is repeated? Also, you can get a better refactoring if you just leave the unmarshalling in the switch, and move the error checking and coverage update to after that.

Reply:
aggh sorry. Yes fixed that and improved refactoring.